ProjFS.Mac/PrjFSKext/FileSystemMountPointsPrivate.hpp

```diff
@@ -0,0 +1,20 @@
+#pragma once
```

Comment:
Can you elaborate on the file name convention?  Previously we've used "Testable" to facilitate testing.

Reply:
The problem I had here is that I needed to access the contents of `struct MountPoint` in the tests in order to prepare the test case. `MountPoint` should however be opaque to the rest of the kext - this is why it was originally defined in `FileSystemMountPoints.cpp`, not the header file which is included by `VirtualizationRoots.cpp` as well.

We explicitly did *not* want the kext code to depend on anything defined in the "…Testable.hpp" headers, i.e. the tail wagging the dog. So if not the main `FileSystemMountPoints.hpp` header, and not `FileSystemMountPointsTestable.hpp`, then there must be some new header that declares and defines things that are private to the implementation but also accessible to tests. Hence `FileSystemMountPointsPrivate.hpp`.

Right solution? Discuss!

Reply:
> So if not the main `FileSystemMountPoints.hpp` header, and not `FileSystemMountPointsTestable.hpp`, then there must be some new header that declares and defines things that are private to the implementation but also accessible to tests. Hence `FileSystemMountPointsPrivate.hpp`.

> Right solution yes/no? Discuss!

When we ran into this issue writing unit tests on Windows that needed access to private ProjFS structs the approach we took was to duplicate the struct somewhere that the test could find it.

@pmj @jeschu1 what do you think of taking the same approach here?  It obviously comes at the cost of maintaining the same struct in two spots, but it comes at the benefit of ensuring the `MountPoint` internals don't leak out to the rest of the kext (and reduces how much we have to contort the code to make it testable).

I also expect that if the `MountPoint` struct is getting updated the unit tests should/will need to get updated anyway, so (in my opinion) it doesn't seem like too heavy of a burden to have a second declaration of the struct for the unit tests.

If we *really* wanted to get fancy we could have a pre-process step that auto-generates a header with `MountPoint` for the unit tests to consume (based on metadata we include in `FileSystemMountPoints.cpp`), but unless maintaining two copies of the struct becomes too much of a hassle that feels like overkill to me.

Reply:
> When we ran into this issue writing unit functional tests on Windows that needed access to private ProjFS structs the approach we took was to duplicate the struct somewhere that the test could find it.
> 
> @pmj @jeschu1 what do you think of taking the same approach here? It obviously comes at the cost of maintaining the same struct in two spots, but it comes at the benefit of ensuring the `MountPoint` internals don't leak out to the rest of the kext (and reduces how much we have to contort the code to make it testable).

I'm not a fan of duplicating struct definitions at all. It's fragile and leads to weird issues that can lead to time wasted head-scratching even if the root cause is discovered before any code gets committed/pushed.

If you're that worried that someone might accidentally `#include` a "forbidden" `…Private.hpp` file, we could add a build step that uses `grep` or whatever to search for them and fails the build.

> One other point, I don't think the function declarations (e.g. `KEXT_STATIC ssize_t FindMountPoint_Locked(mount_t mountPoint);`) should be in a "Private" file, those should be able to stay in a "Testable" file.

Yes, we can move those to stay 100% consistent.

Reply:
Closing the loop on this, as we discussed offline we can stick with the "Private" file approach to avoid duplication.  Thanks!